0012-unify-pulse-commands-and-instructions.md

```diff
@@ -0,0 +1,334 @@
+
+The `Command` will be deprecated. A `Pulse` type will be defined and all pulses will inherit from this.
+
+`Command`s and `Instruction`s will be combined into a new `Instruction` as shown below:
+
+| `Command`      | old `Instruction`         | new `Instruction` | new Operands                                                                                                                 |
+| -------------- | ------------------------- | ----------------- | ---------------------------------------------------------------------------------------------------------------------------- |
+| `PulseCommand` | `PulseInstruction`        | `Play`            | `pulse:Pulse, channel: PulseChannel`                                                                                         |
+| `Delay`        | `DelayInstruction`        | `Delay`           | `duration: int, channel: Channel`                                                                                            |
+| `FrameChange`  | `FrameChangeInstruction`  | `ShiftPhase`      | `phase: float, channel: PulseChannel`                                                                                        |
+| N/A            | N/A                       | `SetPhase`        | `phase: float, channel: PulseChannel`                                                                                        |
+| N/A            | N/A                       | `ShiftFrequency`  | `frequency: float, channel: PulseChannel`                                                                                    |
+| `SetFrequency` | `SetFrequencyInstruction` | `SetFrequency`    | `frequency: float, channel: PulseChannel`                                                                                    |
+| `Acquire`      | `AcquireInstruction`      | `Acquire`         | `duration: int, register: Union[MemorySlot, RegisterSlot], kernel: Optional[Kernel], discriminator: Optional[Discriminator]` |
+| `Snapshot`     | `Snapshot`                | `Snapshot`        | `label: str, snapshot_type: str`                                                                                             |
+
+### Example of pulse programming with changes
+```python
+x90 = gaussian(duration, amp, sigma)
+meas_stim = gaussian_square(duration, amp, sigma, width)
+
+sched = Schedule()
+# we will keep callable shorthand for pulses
+sched += x90(DriveChannel(0))
+sched += Delay(100, DriveChannel(0))
+# or we can generate a Play instruction explicitly
+sched += Play(x90, DriveChannel(0))
+# measurements may be scheduled with delay
+sched += Delay(200, AcquireChannel(0))
+sched += Play(meas_stim, MeasureChannel(0))
+# or shifting in time
+sched += Acquire(100, AcquireChannel(0), MemorySlot(0)) << 200
+# We also have instructions to modify frequency and phase
+sched += SetFrequency(qubit_freq, DriveChannel(0))
```

Comment:
Why use a new DriveChannel instance every time? It looks like you mean to be applying these Instructions to the same DriveChannel - but you're actually creating distinct DriveChannel instances every time. From a performance perspective just passing a reference would avoid a lot of the overhead involved in DriveChannel:__init__. 

Looking at the Instruction(ABC) it appears that each instruction keeps references to its `operands` - thus in your example each Play instruction will be keeping a reference to a distinct DriveChannel object when they all actually represent the same channel.

Why not do this?
```
shed = Schedule()
d0 = DriveChannel(0)
sched += x90(d0)
sched += Delay(100, d0)
...
```

Reply:
This is entirely possible, I'm just avoiding defining additional references for the sake of these examples.

Reply:
fair enough